source/postprocess/particle_crystal_preferred_orientation.cc

```diff
@@ -0,0 +1,1036 @@
+      const double phi2 = phi2_d *degree_to_rad;
+      Tensor<2,3> rot_matrix;
+
+
+      rot_matrix[0][0] = cos(phi2)*cos(phi1) - cos(theta)*sin(phi1)*sin(phi2);
+      rot_matrix[0][1] = -cos(phi2)*sin(phi1) - cos(theta)*cos(phi1)*sin(phi2);
+      rot_matrix[0][2] = -sin(phi2)*sin(theta);
+
+      rot_matrix[1][0] = sin(phi2)*cos(phi1) + cos(theta)*sin(phi1)*cos(phi2);
+      rot_matrix[1][1] = -sin(phi2)*sin(phi1) + cos(theta)*cos(phi1)*cos(phi2);
+      rot_matrix[1][2] = cos(phi2)*sin(theta);
+
+      rot_matrix[2][0] = -sin(theta)*sin(phi1);
+      rot_matrix[2][1] = -sin(theta)*cos(phi1);
+      rot_matrix[2][2] = cos(theta);
+      AssertThrow(rot_matrix[2][2] <= 1.0, ExcMessage("rot_matrix[2][2] > 1.0"));
+      return rot_matrix;
+    }
+
+
+
+    template <int dim>
+    void
+    ParticleCrystalPreferredOrientation<dim>::set_last_output_time (const double current_time)
+    {
+      // if output_interval is positive, then update the last supposed output
+      // time
+      if (output_interval > 0)
+        {
+          // We need to find the last time output was supposed to be written.
+          // this is the last_output_time plus the largest positive multiple
+          // of output_intervals that passed since then. We need to handle the
+          // edge case where last_output_time+output_interval==current_time,
+          // we did an output and std::floor sadly rounds to zero. This is done
+          // by forcing std::floor to round 1.0-eps to 1.0.
+          const double magic = 1.0+2.0*std::numeric_limits<double>::epsilon();
+          last_output_time = last_output_time + std::floor((current_time-last_output_time)/output_interval*magic) * output_interval/magic;
+        }
+    }
+
+    template <int dim>
+    typename ParticleCrystalPreferredOrientation<dim>::Output
+    ParticleCrystalPreferredOrientation<dim>::string_to_output_enum(std::string string)
+    {
+      //olivine volume fraction, olivine A matrix, olivine Euler angles, enstatite volume fraction, enstatite A matrix, enstatite Euler angles
+      if (string == "volume fraction")
+        return Output::VolumeFraction;
+      if (string == "rotation matrix")
+        return Output::RotationMatrix;
+      if (string == "Euler angles")
+        return Output::EulerAngles;
+      return Output::not_found;
+    }
+
+
+    template <int dim>
+    void
+    ParticleCrystalPreferredOrientation<dim>::declare_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection("Postprocess");
+      {
+        prm.enter_subsection("Particles");
+        {
+          prm.enter_subsection("Crystal Preferred Orientation");
+          {
+            prm.declare_entry ("Number of grains per praticle", "50",
+                               Patterns::Integer (0),
+                               "The number of grains of olivine and the number of grain of enstatite "
+                               "each particle contains.");
```

Comment:
You dont need to declare this again, if it is already declared in the particle property. Same below with the other parameters.